cmd/azureBlobUpload.go

```diff
@@ -0,0 +1,120 @@
+// We introduce this 'wrapper' for mocking reasons
+func UploadFile(ctx context.Context, api *azblob.BlockBlobClient, file *os.File, o azblob.HighLevelUploadToBlockBlobOption) (*http.Response, error) {
+	return api.UploadFileToBlockBlob(ctx, file, o)
+}
+
+// Struct to store Azure credentials from specified JSON string
+type azureCredentials struct {
+	SASToken    string `json:"sas_token"`
+	AccountName string `json:"account_name"`
+	Container   string `json:"container_name"`
+}
+
+func azureBlobUpload(config azureBlobUploadOptions, telemetryData *telemetry.CustomData) {
+	// Prepare Credentials
+	log.Entry().Infoln("Start reading Azure Credentials")
+	var obj azureCredentials
+
+	err := json.Unmarshal([]byte(config.JSONCredentialsAzure), &obj)
+	if err != nil {
+		log.Entry().
+			WithError(err).
+			Fatal("Could not read JSONCredentialsAzure")
+	}
+
+	// Initialize Azure Service Client
+	sasURL := fmt.Sprintf("https://%s.blob.core.windows.net/?%s", obj.AccountName, obj.SASToken)
+	serviceClient, err := azblob.NewServiceClientWithNoCredential(sasURL, nil)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("Could not instantiate Azure Service Client!")
+	}
+
+	// Get a containerClient from ServiceClient
+	containerClient := serviceClient.NewContainerClient(obj.Container)
+
+	err = runAzureBlobUpload(&config, containerClient, UploadFile)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("step execution failed")
+	}
+}
+
+func runAzureBlobUpload(config *azureBlobUploadOptions, containerClient AzureContainerAPI, Upload func(ctx context.Context, api *azblob.BlockBlobClient, file *os.File, o azblob.HighLevelUploadToBlockBlobOption) (*http.Response, error)) error {
+
+	log.Entry().Infof("Start walk through FilePath '%v'", config.FilePath)
+
+	// All Blob Operations operate with context.Context, in our case the clients do not expire
+	ctx := context.Background()
+
+	//iterate through directories
+	err := filepath.Walk(config.FilePath, func(currentFilePath string, f os.FileInfo, err error) error {
+		// Handle Failure to prevent panic (e.g. in case of an invalid filepath)
+		if err != nil {
+			log.Entry().WithError(err).Warnf("Failed to access path: '%v'", currentFilePath)
+			return err
+		}
+		// Skip directories, only upload files
+		if !f.IsDir() {
+			log.Entry().Infof("Current target path is: '%v'", currentFilePath)
+
+			// Read Data from File
+			data, e := os.Open(currentFilePath)
+			if e != nil {
+				log.Entry().WithError(e).Warnf("Could not read the file '%s'", currentFilePath)
+				return e
+			}
+			defer data.Close()
+
+			// Create a filepath in UNIX format so that the BlockBlobClient automatically detects directories
+			key := filepath.ToSlash(currentFilePath)
+
+			// Get a blockBlobClient from containerClient
+			blockBlobClient := NewBlockBlobClient(key, containerClient)
+
+			// Upload File
+			log.Entry().Infof("Start upload of file '%v'", currentFilePath)
+			var blockOptions azblob.HighLevelUploadToBlockBlobOption
```

Comment:
This looks empty, but it is used in the next line for the Upload. Do we need this at all?

Reply:
We need a struct of type "HighLevelUploadToBlockBlobOption" for the (azblob.BlockBlobClient).UploadFileToBlockBlob function. Since I have not given any further specifications, e.g. for the "BlobSize", it is, as you have correctly recognised, an empty struct. [In this code example under "Upload blobs to the container" it is handled in the same way.](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-quickstart-blobs-go)

Reply:
ok